src/pymor/reductors/mt.py

```diff
@@ -0,0 +1,210 @@
+        assert which in ('LR', 'SM', 'NR', 'NS', 'NM')
+        assert method_options is None or isinstance(method_options, dict)
+        if not method_options:
+            method_options = {}
+
+        if self.fom.parametric:
+            fom = self.fom.with_(**{op: getattr(self.fom, op).assemble(mu=self.mu)
+                                    for op in ['A', 'B', 'C', 'D', 'E']})
+        else:
+            fom = self.fom
+
+        self.V = fom.A.source.empty(reserve=r)
+        self.W = fom.A.source.empty(reserve=r)
+
+        if decomposition == 'eig':
+            if fom.order >= sparse_min_size():
+                if not isinstance(fom.A, NumpyMatrixOperator) or fom.A.sparse:
+                    self.logger.warning('Converting operator A to a NumPy array.')
+                if not isinstance(fom.E, IdentityOperator):
+                    if not isinstance(fom.E, NumpyMatrixOperator) or fom.E.sparse:
+                        self.logger.warning('Converting operator E to a NumPy array.')
+            A = to_matrix(fom.A, format='dense')
+            E = None if isinstance(fom.E, IdentityOperator) else to_matrix(fom.E, format='dense')
+
+            if symmetric:
+                poles, ev_r = spla.eig(A, E, right=True)
+                rev = fom.A.source.from_numpy(ev_r.T)
+                lev = rev.copy()
+            else:
+                poles, ev_l, ev_r = spla.eig(A, E, left=True, right=True)
+                rev = fom.A.source.from_numpy(ev_r.T)
+                lev = fom.A.source.from_numpy(ev_l.T)
+            if which == 'SM':
+                idx = np.argsort(np.abs(poles))
+            elif which == 'LR':
+                idx = np.argsort(-poles.real)
+            else:
+                absres = np.empty(len(poles))
+                for i in range(len(poles)):
+                    lev[i].scal(1 / lev[i].inner(fom.E.apply(rev[i]))[0][0])
+                    b_norm = fom.B.apply_adjoint(lev[i]).l2_norm()
+                    c_norm = fom.C.apply(rev[i]).l2_norm()
+                    absres[i] = b_norm @ c_norm
+                if which == 'NR':
+                    idx = np.argsort(-absres / np.abs(np.real(poles)), kind='stable')
+                elif which == 'NS':
+                    idx = np.argsort(-absres / np.abs(poles))
+                elif which == 'NM':
+                    idx = np.argsort(-absres)
+            poles = poles[idx]
+            rev = rev[idx]
+            lev = lev[idx]
+
+        elif decomposition == 'samdp':
+            poles, res, rev, lev = samdp(fom.A, fom.E,
+                                         fom.B.as_range_array(),
+                                         fom.C.as_source_array(),
+                                         r, **method_options)
+        if np.iscomplexobj(poles):
+            real_index = np.where(np.isreal(poles))[0]
+            complex_index = np.where(poles.imag > 0)[0]
+            if len(np.where(poles[:r].imag)[0]) % 2 == 1:
```

Comment:
I guess this assumes that the conjugate pairs in `poles` are always next to each other. I'm not sure this is true when there are two pairs with the same real parts, e.g.,

```python
>>> import numpy as np
>>> a = np.array([1j, -1j, 2j, -2j])
>>> a
array([ 0.+1.j, -0.-1.j,  0.+2.j, -0.-2.j])
>>> a.sort()
>>> a
array([-0.-2.j, -0.-1.j,  0.+1.j,  0.+2.j])
```

It might be good to test the algorithm on the FOM example from the [SLICOT benchmark collection](http://slicot.org/20-site/126-benchmark-examples-for-model-reduction).

Reply:
Not really. Or respectively the construction of the projection matrices takes care of it. So when constructing the projection matrices V and W the real and imaginary part of the eigenvector from the first complex pole that occurs are taken. Such that for the example above the reduced model would contain the complex pair (2j, -2j) even though the first two poles after sorting are -2j and -1j. On the FOM example it works as expected and gives a warning when reducing to order 3 for example.

Reply:
I'm thinking more about the example where `r` is 2. I guess then this `if` branch thinks that `-2j` and `-1j` form a conjugate pair.

Reply:
True, but the construction later on takes the real and imaginary part of the eigenvector of `-2j`  such that `-1j` would not be a pole of the reduced model.

Reply:
Hm, how? If `r` is 2, then the number of complex numbers is even, so the execution goes to the `else` part where `complex_index = np.where(poles.imag > 0)[0]`, right?

Reply:
Right, but yeah now I see. The selection of the complex_index needs to be the same as in the else part if the pair is split. So not only checking if `poles.imag >0` but also `poles.imag <0` and taking the index list which is longer. With that changed it should work, right?

Reply:
I'm not sure what you mean. Both `poles.imag > 0` and `poles.imag < 0` should be of the same length. I guess you mean `poles[:r].imag`?
Either way, the condition in this `if` branch (`len(np.where(poles[:r].imag)[0]) % 2 == 1`) should not assume that `poles[:r]` are the "right" poles. Maybe you can look at `poles[poles.imag >= 0]` and count the first `r` poles while double-counting the complex poles, e.g., with something like `np.cumsum((poles[poles.imag >= 0].imag > 0).astype(int) + 1)`.

Reply:
Right. 
In my opinion that would make this condition more complicated than necessary since one would still need to look at `poles.imag[:r] > 0` and `poles.imag[:r] < 0` when constructing the projection matrices. So my suggestion is to use the same construction as in lines 164-178 in the else branch of this if statement.

                    real_index = np.where(np.isreal(poles))[0]
                    complex_index = np.where(poles.imag > 0)[0] \
                        if np.sum(np.where(poles.imag > 0)) > np.sum(np.where(poles.imag < 0)) \
                        else np.where(poles.imag < 0)[0]

                    self.V.append(rev[real_index].real)
                    self.V.append(rev[complex_index].real)
                    self.V.append(rev[complex_index].imag)

                    self.W.append(lev[real_index].real)
                    self.W.append(lev[complex_index].real)
                    self.W.append(lev[complex_index].imag)

                    self.V = self.V[:r]
                    self.W = self.W[:r]

With that the if condition would still assume that `poles[:r]` are the "right" poles but I don't see why this is a problem as long as the poles or more the eigenvectors taken for the projection are the right ones. 

Another idea I had was to use stable sorting algorithms to make sure that the poles would come in pairs but I guess that would make things less efficient.

Reply:
I'm not sure the above approach chooses the "right" eigenvectors. As far as I see, for `poles = [-2j, -1j, 1j, 2j]` and `r = 2`, it would only take the real parts of the eigenvectors corresponding to `-2j` and `-1j`, and it should take the real and imaginary part of one of those. For `poles = [-3j, -2j, -1j, 1j, 2j, 3j]` and `r = 3`, it's similar, but it should take the real and imaginary part from one complex pair and something from another pair.

A stable sorting algorithm would work if `poles` starts out with complex pairs next to each other. That might be true for `eig`, but I'm not sure it is for `samdp`. (Efficiency matters less for reduced quantities.) You could use something like

```python
idx = sorted(
    range(len(poles)),
    key=lambda i: (-dominance[i],
                   -poles[i].real,
                   abs(poles[i].imag),
                   0 if poles[i].imag >= 0 else 1))
poles = poles[idx]
```

assuming `dominance[i] == dominance[j]` when `poles[i]` and `poles[j]` form a complex conjugate pair (or this could be ensured in the code).

Reply:
Ah, you're absolutely right. I was a fool, sorry for that. 
So one could also force the sorting in the `samdp` algorithm to use a stable algorithm and do the same for the sorting after the use of the `eig` algorithm. Or is there a reason why the sorting in samdp is not stable @lbalicki? Then the complex pairs would be next to each other, might be the easiest way.

Reply:
Well, you'll need this special sorting whether `samdp` does it or not since it's needed for the `eig` approach (assuming you go for the sorting approach). It might be good for `samdp` to do this sorting, but I guess it's not strictly necessary.